### PR TITLE
GEODE-6762: First part - check is there proxy to notify

### PR DESCRIPTION
**Pull request by mivanac**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [*] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [*] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [*] Is your initial contribution a single, squashed commit?

- [*] Does `gradlew build` run cleanly?

- [*] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientNotifier.java [hunk 1]
```diff
@@ -599,6 +599,18 @@ public static void notifyClients(InternalCacheEvent event) {
...
+  public static boolean singletonHasClientProxies() {
+    CacheClientNotifier instance = ccnSingleton;
+    if (instance != null) {
+      return instance.hasClientProxies();
+    }
+    return false;
+  }
+
+  private boolean hasClientProxies() {
+    return !this._clientProxies.isEmpty() || !this._initClientProxies.isEmpty();
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientNotifier.java]: I think there a possible race condition where we check the first isEmpty() method, thread is interrupted, then in the different thread we move a clientProxy from initClientProxies into the clientProxies map.  We now resume the first thread and  check initClientProxies and this is also now empty, but we really had 1 client proxy the entire time?

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientNotifier.java]: I think you are right. When we promote a proxy from the init map to the clientProxies map, we add it to clientProxies first and then remove it from init. So it is in both maps for a bit.
So I think this code can be fixed by checking the init map first. Does this sound correct?

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/CacheClientNotifierTest.java [hunk 2]
```diff
@@ -175,4 +177,40 @@ private InternalCacheEvent createMockInternalCacheEvent(
...
+    CacheClientNotifier ccn =
+        CacheClientNotifier.getInstance(internalCache, mock(CacheServerStats.class), 10,
+            10, mock(ConnectionListener.class), null, true);
+
+    assertFalse(CacheClientNotifier.singletonHasClientProxies());
+    ccn.shutdown(111);
+
+  }
+
+  @Test
```
Comment (dschneider-pivotal) [inline geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/CacheClientNotifierTest.java]: Should you also test that addClientInitProxy causes SingletonHasClientProxiesTrue?

Comment (mivanac) [reply geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/CacheClientNotifierTest.java]: Thanx for comments. I will update PR.

**Comment by jhuynh1** [comment]
@nabarunnag would you be able to review in case I missed anything else?

**Comment by nabarunnag** [comment]
Changes look good with initial testing. Merging the PR and monitoring if there are unknown effects.

**Pull request merged**